### DISCUSSION TITLE
Need your help investigating memory leak on .Net Framework

### DISCUSSION THREAD

**Discussion by En3Tho** [dotnet/runtime#99111] [Q&A]
We have a service exhibiting a memory leak in a weird way: seems like GC refuses to do gen2 collections and reasons to that are unknown.

So gen2 keeps growing, but I don't see GC even trying to clean up gen2. We've taken few ~30sec PerfView traces and not one of them reported a gen2 collection.

An example of PerfView report:

| Gen | Count | Max Pause | Max Peak MB | Max Alloc MB/sec | Total Pause | Total Alloc MB | Alloc MB/MSec GC | Survived MB/MSec GC | Mean Pause | Induced |
| --- | ----- | --------- | ----------- | ---------------- | ----------- | -------------- | ---------------- | ------------------- | ---------- | ------- |
| ALL | 119   | 28,6      | 9 977,2     | 1 349,902        | 696,3       | 65 554,3       | 94,1             | 0,092               | 5,9        | 0       |
| 0   | 64    | 11,6      | 9 968,2     | 1 349,902        | 339,6       | 34 621,8       | 193,0            | 0,079               | 5,3        | 0       |
| 1   | 55    | 28,6      | 9 977,2     | 1 271,659        | 356,7       | 30 932,5       | 183,8            | 0,105               | 6,5        | 0       |
| 2   | 0     | 0,0       | 0,0         | 0,000            | 0,0         | 0,0            | ∞                | NAN                 | NAN        | 0       |

I think the reason for the leak is a constant GCHandle grow. So far with heap of 10 gigbytes (8.2 of which is gen2) there are 7.3 million GCHandles reported. Looking for `clr!GCHandleStore::CreateHandleOfType` wasn't fruitful because callers are knowns framework functions.

I need some help on how to diagnose this kind of issue. The weirdest thing to me, again, is that gc refuses to make gen2 collections. Every gen0/gen1 collection ends up cleaning 300 to 500 megabytes of data and it seems fine with just that. I'm confused whether this is a leak after all or just gc refusing to clean up these handles. Note that there are other services with the same verision/dlls and they do not behave like that.

**Comment by En3Tho** [comment]
Can you please take a look at this please, @Maoni0 ?

**Comment by SerratedSharp** [comment]
Memory leaks in .NET are often due to some pattern in how your objects are related to each other (which can also include those through delegates/event) that keeps objects referenced despite perhaps conceptually not being needed.  Misconfigured DI containers can also cause an explosion in memory usage.  Have you looked at what types are the biggest consumers?
<img width="474" alt="image" src="https://github.com/dotnet/runtime/assets/97156524/5b280d47-f088-491f-a33c-36755aaefcee">
